src/shared.ts

```diff
@@ -90,3 +96,24 @@ export async function getEnterpriseConnections(context: CacheContext) {
 		return enterpriseConnections;
 	});
 }
+
+export const DefaultCacheTimeMinutes = 30;
+
+export const sessionCachedFetch = async <T>(
+	key: SessionCacheKey,
+	cacheTimeMinutes: number,
+	fetchFn: () => Promise<T>,
+) => {
+	const sessionStorage = await storage.session.get(key);
+	const data = sessionStorage[key] as CachedFetchResponse<T> | undefined;
+	if (data && data.timestamp > Date.now() - cacheTimeMinutes * 60 * 1000) {
+		return data.data;
+	}
+
+	const newData = await fetchFn();
+	if (newData) {
+		await storage.session.set({ [key]: { data: newData, timestamp: Date.now() } });
+	}
+
+	return newData;
```

Comment:
Should we still call the fetch if we get a cache, hit, just not wait for it to return the cached data?

Reply:
for rate limit reasons I want to say, not immediately. we basically want to introduce the concept of cache time vs stale time (something that react-query) has, so it fetches in the background only if the state time has passed. I want to do make caching a lot better in a followup but I currently have another branch on top of this so I'll revisit caching after that branch(es) get merged.